service/interpreter/workflowImpl.go

```diff
@@ -763,28 +765,35 @@ func invokeStateExecute(
 		// this is not a problem because the signalWithStart will be very fast and highly available
 		unifiedClient := env.GetUnifiedClient()
 
-		var workflowId string
+		signalWithStartOn := config.GetSignalWithStartOnWithDefault(env.GetSharedConfig())
 
-		if state.WaitForKey != nil {
-			workflowId = service.IwfSystemConstPrefix + executionContext.WorkflowId + "_" + state.StateId + "_" + *state.WaitForKey
-		} else {
-			workflowId = service.IwfSystemConstPrefix + executionContext.WorkflowId + "_" + *executionContext.StateExecutionId
+		// signalWithStart with legacy workflowId (containing parent workflowId)
+		if provider.GetBackendType() == service.BackendTypeCadence ||
+			(provider.GetBackendType() == service.BackendTypeTemporal && (signalWithStartOn == "old" || signalWithStartOn == "both")) {
+			workflowId := utils.GetWorkflowIdForWaitForStateExecution(executionContext.WorkflowId, executionContext.StateExecutionId, state.WaitForKey, &state.StateId)
+
+			err = signalWithStart(unifiedClient, workflowId)
+			if err != nil && !unifiedClient.IsWorkflowAlreadyStartedError(err) {
+				// WorkflowAlreadyStartedError is returned when the started workflow is closed and the signal is not sent
+				// panic will let the workflow task will retry until the signal is sent
+				panic(fmt.Errorf("failed to signal on completion %w", err))
```

Comment:
Nit: it will look slightly better to move this error handling into the signalWithStart method

Reply:
Addressed